### ISSUE TITLE
ArrayIndexOutOfBoundsException on ConcurrentDoubleHistogram#recordValue

### ISSUE THREAD

**Issue by rrakos-evo** [HdrHistogram/HdrHistogram#156]
I'm working on implementing a `HdrHistogram` based `Summary` metric collector for Prometheus, as the default `CKMS` based `Summary` metric collector is not performant enough for our use cases.
https://github.com/prometheus/client_java/pull/484

We experienced `ArrayIndexOutOfBoundsException` on `ConcurrentDoubleHistogram#recordValue` during load testing under realistic workload(s) with high-throughput and high-concurrency (200mn+ calls / day, 100+ threads).
I couldn't yet reproduce the same exception under controlled conditions (`-ea`, unit tests, JMH benchmarks, serial tests, concurrent tests, static data, randomized data, ...).
I could only trigger similar exceptions when a measurement is outside of the widest possible dynamic range, but that's a different scenario which is expected for invalid measurements.
https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/DoubleHistogram.java#L557
https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/DoubleHistogram.java#L416

It looks like it’s a rare concurrency bug happening when recording values while `auto-resizing` or `value shifting`, but I'm still investigating and exploring `HdrHistogram` code to understand the details.

1st occurrence - immediately after restart, no error handling:
```
java.lang.ArrayIndexOutOfBoundsException: Index 67584 out of bounds for length 65536
  at java.base/java.lang.invoke.VarHandle$1.apply(VarHandle.java:2011)
  at java.base/java.lang.invoke.VarHandle$1.apply(VarHandle.java:2008)
  at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:159)
  at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:156)
  at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:62)
  at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
  at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
  at java.base/java.lang.invoke.VarHandleLongs$Array.getAndAdd(VarHandleLongs.java:721)
  at java.base/java.lang.invoke.VarHandleGuards.guard_LIJ_J(VarHandleGuards.java:778)
  at java.base/java.util.concurrent.atomic.AtomicLongArray.incrementAndGet(AtomicLongArray.java:234)
  at org.HdrHistogram.ConcurrentHistogram.recordConvertedDoubleValue(ConcurrentHistogram.java:169)
  at org.HdrHistogram.DoubleHistogram.recordSingleValue(DoubleHistogram.java:353)
  at org.HdrHistogram.DoubleHistogram.recordValue(DoubleHistogram.java:294)
  at io.prometheus.client.TimeWindowQuantiles.insert(TimeWindowQuantiles.java:53)
  at io.prometheus.client.Summary$Child.observe(Summary.java:264)
  ...
```

2nd occurrence - during normal operation, improved error handling:
```
java.lang.RuntimeException: Failed to record 0.019 in bucket org.HdrHistogram.ConcurrentDoubleHistogram@47729061
encodeIntoCompressedByteBuffer: DHISTwAAAAMAAAAAAAAABByEkxQAAABBeNqTaZkszMDAoMwAAcxQmhFC2f+3OwBhHZdgecrONJWDpZuF5zozSzMTUz8bVzcL03VmjkZGlnqWRkY2AGoTC78=
  at io.prometheus.client.TimeWindowQuantiles.insert(TimeWindowQuantiles.java:65)
  at io.prometheus.client.Summary$Child.observe(Summary.java:264)
  ...
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 4317 out of bounds for length 4096
  at java.base/java.lang.invoke.VarHandle$1.apply(VarHandle.java:2011)
  at java.base/java.lang.invoke.VarHandle$1.apply(VarHandle.java:2008)
  at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:159)
  at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:156)
  at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:62)
  at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
  at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
  at java.base/java.lang.invoke.VarHandleLongs$Array.getAndAdd(VarHandleLongs.java:721)
  at java.base/java.lang.invoke.VarHandleGuards.guard_LIJ_J(VarHandleGuards.java:778)
  at java.base/java.util.concurrent.atomic.AtomicLongArray.incrementAndGet(AtomicLongArray.java:234)
  at org.HdrHistogram.ConcurrentHistogram.recordConvertedDoubleValue(ConcurrentHistogram.java:169)
  at org.HdrHistogram.DoubleHistogram.recordSingleValue(DoubleHistogram.java:353)
  at org.HdrHistogram.DoubleHistogram.recordValue(DoubleHistogram.java:294)
  at io.prometheus.client.TimeWindowQuantiles.insert(TimeWindowQuantiles.java:57)
  ...
```

3rd occurrence - during normal operation, improved error handling:
```
java.lang.RuntimeException: Failed to record 0.012 in bucket org.HdrHistogram.ConcurrentDoubleHistogram@ed824791
encodeIntoCompressedByteBuffer: DHISTwAAAAMAAAAAAAAAAhyEkxQAAAAieNqTaZkszMDAwMIAAcxQmhFCyf+32wBhMa0VYAIAUp8EHA==
  at io.prometheus.client.TimeWindowQuantiles.insert(TimeWindowQuantiles.java:65)
  at io.prometheus.client.Summary$Child.observe(Summary.java:264)
  ...
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 4644 out of bounds for length 4096
  at java.base/java.lang.invoke.VarHandle$1.apply(VarHandle.java:2011)
  at java.base/java.lang.invoke.VarHandle$1.apply(VarHandle.java:2008)
  at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:159)
  at java.base/jdk.internal.util.Preconditions$1.apply(Preconditions.java:156)
  at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:62)
  at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
  at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
  at java.base/java.lang.invoke.VarHandleLongs$Array.getAndAdd(VarHandleLongs.java:721)
  at java.base/java.lang.invoke.VarHandleGuards.guard_LIJ_J(VarHandleGuards.java:778)
  at java.base/java.util.concurrent.atomic.AtomicLongArray.incrementAndGet(AtomicLongArray.java:234)
  at org.HdrHistogram.ConcurrentHistogram.recordConvertedDoubleValue(ConcurrentHistogram.java:169)
  at org.HdrHistogram.DoubleHistogram.recordSingleValue(DoubleHistogram.java:353)
  at org.HdrHistogram.DoubleHistogram.recordValue(DoubleHistogram.java:294)
  at io.prometheus.client.TimeWindowQuantiles.insert(TimeWindowQuantiles.java:57)
  ...
```

I improved error handling by serializing the histogram and logging the offending value: 
```
    try {
        bucket.recordValue(value);
    } catch (ArrayIndexOutOfBoundsException e) {
        ByteBuffer byteBuffer = ByteBuffer.allocate(bucket.getNeededByteBufferCapacity());
        bucket.encodeIntoCompressedByteBuffer(byteBuffer, Deflater.BEST_COMPRESSION);
        byteBuffer.flip();
        byte[] byteArray = new byte[byteBuffer.limit()];
        byteBuffer.get(byteArray);
        String base64 = Base64.getEncoder().encodeToString(byteArray);
        throw new RuntimeException("Failed to record " + value + " in bucket " + bucket + "\n" + "encodeIntoCompressedByteBuffer: " + base64, e);
    }
```

I verified that the deserialized histogram is not corrupted and the value could be recorded:
```
    double value = ...;
    String base64 = ...;

    byte[] byteArray = Base64.getDecoder().decode(base64);
    ByteBuffer byteBuffer = ByteBuffer.wrap(byteArray);
    DoubleHistogram bucket = ConcurrentDoubleHistogram.decodeFromCompressedByteBuffer(byteBuffer, ConcurrentHistogram.class, 2);
    bucket.setAutoResize(true);

    bucket.recordValue(value);
```

It looks like increasing the initial dynamic range avoids the issue as it reduces `auto-resizing` or `value shifting`, but it might still fail on larger measurements which trigger `auto-resizing` or `value shifting`.
https://github.com/prometheus/client_java/pull/484#discussion_r299942012

We could swallow the exception and drop the measurement, so this is not a showstopper for our use cases, but I'm not sure if it's acceptable for others, especially use cases other than application metrics.

**Comment by giltene** [comment]
The improved error reporting, reporting the compressed histogram with the error) is very useful. I may start doing that in the core code in order to get better debugging info fir cases like this, I’ll dig into the two reported histograms above and update you here.

**Comment by giltene** [comment]
I think I've found an explanation, and fixed the issue with https://github.com/HdrHistogram/HdrHistogram/commit/753fe5c16ddb089b42a6ff920c06e50ead032945 .

The captured compressed histograms above were very useful. The second one only had two values recorded and was still at the original size (4096 integer array element), which really helped narrow down the possibilities. Fundamentally, non-resizing shifts [which can occur when there are empty parts of the histogram] can make a value that was valid prior to the shift invalid after the shift, and under a race. Luckily, the number of such shifts is capped at a theoretical 64 for the lifetime of a histogram, so a retry-based fix is straightforward.

Unfortunately, since I don't have a reproducer, I have not yet verified the fix. _**Would you be able to build and test this in your environment under the sort of load that had been producing the problem before?**_

BTW, I see [from the captured histograms] you use 3 decimal point resolution, which is fine, but I've found that 2 decimal points tend to be plenty from a precision point of view for the vast majority of applications (and the underlying storage requirements are ~8x lower). Does the use case really need 3 decimal points?

**Comment by rrakos-evo** [comment]
Thank you for the extremely quick turnaround. 🙂

I'm happy to hear that the captured histograms were useful. I was a bit worried that they could be misleading because of concurrent updates changing their state meanwhile taking the snapshots.

I was quite confused when I was investigating, because if I understand correctly, the active and inactive arrays can't shrink and the index is normalized and validated, yet it was was still causing an AIOOBE. I'm quite sure I was looking at the wrong place, as I don't understand the implementation details enough.
https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/ConcurrentHistogram.java#L605
https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/AbstractHistogram.java#L2304

I'll prepare a build with the changes and run our load tests over the weekend and next week. I think a couple of days should be enough to verify that the fix resolves the issue without affecting performance.

I'll revert the initial dynamic range back to the default for the tests, as that seems to avoid the issue.
https://github.com/prometheus/client_java/pull/484#discussion_r299942012

I don't think we need 3 decimal points resolution, as it shouldn't matter if a duration or latency is `100ms` vs `101ms` or `100s` vs `101s`. I used the same resolution for `CKMS` and `HdrHistogram` for regression tests and I left it the same for benchmarks and load tests. I think we still need to find a balance between memory usage vs value resolution and initial dynamic range vs resize performance overhead, especially that there might be many histograms around because of time windowing and Prometheus metric labels.

**Comment by giltene** [comment]
> Thank you for the extremely quick turnaround. 🙂
> 
> I'm happy to hear that the captured histograms were useful. I was a bit worried that they could be misleading because of concurrent updates changing their state meanwhile taking the snapshots.

The first of the two histograms was captured after the a resize, such that the index that caused the AIOOB was valid at the time of the capture. But the second was captured with a size that would still cause the AIOOB. That was actually one a pretty big hint that saved me a lot of time: look for situations where a resize may not be needed for this bug to occur.

> 
> I was quite confused when I was investigating, because if I understand correctly, the active and inactive arrays can't shrink and the index is normalized and validated, yet it was was still causing an AIOOBE. I'm quite sure I was looking at the wrong place, as I don't understand the implementation details enough.

I was looking at that wrong place too. And at the various resizing code paths. A lot. I (for now) convinced myself that whatever path you go through which expands the array in size can't result in a value that is valid (per the the test in https://github.com/HdrHistogram/HdrHistogram/blob/69064c4cc9837ef8178452eb56aef7184a550c84/src/main/java/org/HdrHistogram/DoubleHistogram.java#L347) becoming invalid later and throwing an AIOOB when trying to record it into the integer histogram. This is because the resizing paths only expand the valid range, and never contract it on either side. Specifically, the range-expanding cases determine the expanded range that here:
https://github.com/HdrHistogram/HdrHistogram/blob/69064c4cc9837ef8178452eb56aef7184a550c84/src/main/java/org/HdrHistogram/DoubleHistogram.java#L465
and here:
https://github.com/HdrHistogram/HdrHistogram/blob/69064c4cc9837ef8178452eb56aef7184a550c84/src/main/java/org/HdrHistogram/DoubleHistogram.java#L536
And they do it in a synchronized fashion (against any other range adjuster) and in a way that would never appear to a concurrent (unsynchronized) observer to "shrink" either end of the valid value "auto range".

That left the "what if the range is not expanded?" paths. If auto-ranging successfully shifts the values, either right or left, without encountering a need to resize (the shift does not overflow and does not stomp existing recorded values) then the valid value range for the double histogram would shifts, expanding on one side and _**contracting**_ on the other. It's the contracting side that causes this bug: a value that passed the (unsynchronized) auto-range check can then end up being recorded into the integer histogram after is was fully and properly shifted. E.g. a shift can both start and complete between the auto range test and the call to integerValuesHistogram.recordConvertedDoubleValueWithCount().

The fix is then obvious: If we see an AIOOB, redo the auto-ranging test and try again. Since the number of auto-ranging operations that can cause such a race is capped at a small value (64) during the lifetime of any one histogram instance, this retry is safe to do without fear of live-locking [but I now test for that just in case].

> https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/ConcurrentHistogram.java#L605
> https://github.com/HdrHistogram/HdrHistogram/blob/master/src/main/java/org/HdrHistogram/AbstractHistogram.java#L2304
> 
> I'll prepare a build with the changes and run our load tests over the weekend and next week. I think a couple of days should be enough to verify that the fix resolves the issue without affecting performance.
> 
> I'll revert the initial dynamic range back to the default for the tests, as that seems to avoid the issue.
> [prometheus/client_java#484 (comment)](https://github.com/prometheus/client_java/pull/484#discussion_r299942012)
> 
> I don't think we need 3 decimal points resolution, as it shouldn't matter if a duration or latency is `100ms` vs `101ms`. I used the same resolution for `CKMS` and `HdrHistogram` for regression tests and I left it the same for benchmarks and load tests. I think we still need to find a balance between memory usage vs value resolution and initial dynamic range vs resize performance overhead, especially that there might be many histograms around because of time windowing and Prometheus metric labels.

Autosizing pre-sized histograms works fine, but the way I convinced myself that keeping the histograms minimally sized is probably the way to go in large systems is that the number of auto-sizing and auto-ranging operations in the lifetime of one histograms is extremely limited (the count of resizing and shifting events can't total more than 64). So you can think of them as lazy-construction costs as opposed to general recording costs. This is the same logic that makes it "ok" (even if not ideal) to synchronize the slow path of the auto-sizing operations in otherwise-concurrent histograms: The total number of blocking recording events that can occur in the life of a histogram is capped at a very small number...

Another question on use case: You are making direct use of the ConcurrentDoubleHistogram here, and whenever I see that, it always makes me wonder about what you do to eventually get the histogram "at rest" in order to read it in some way, and how you prevent additional modifications from happening. The DoubleRecorder was designed for this very common need (and is actually the reason I built ConcurrentDoubleHistogram to begin with). Have you looked at using it?  I've found that since creating the Recorder and DoubleRecorder, I have never personally found a use case where I used ConcurrentHistogram or ConcurrentDoubleHistogram directly. The recorders seem to "always be the right thing" to use in concurrent recording cases. They do double the footprint, but the ability to gain a stable interval histogram to read and work on is usually "worth it", often removing other copies that would have been kept around when trying to achieve "sampling" in concurrent situations in some other way.

**Comment by rrakos-evo** [comment]
I got sidetracked by more urgent things. I'll try to get back to this next week.
I'm sorry about the inconvenience. 🙁

**Comment by rrakos-evo** [comment]
I'm sorry, but I still couldn't get back to this, but I'll try this week. 🙁

**Comment by giltene** [comment]
Any new news?

**Comment by rrakos-evo** [comment]
I believe the fix works as intended, as I don't remember seeing similar exceptions.
I deployed it immediately, and it worked fine for a couple of days, until the environment collapsed.

I'm not entirely happy with the way we're measuring latencies and handling metrics in general.
I'm sorry that I still didn't read your reply (https://github.com/HdrHistogram/HdrHistogram/issues/156#issuecomment-508809060) and I can't yet answer your questions. I'll try again tomorrow. 😞

I spent ~3 days on this task over the past month. 😢 
I spent ~3 weeks fixing our testing environments and investigating issues.
I spent ~1 week helping team members and doing code reviews.

**Comment by rrakos-evo** [comment]
I'm sorry this has been taking so long, but I'm coming back with good news.
We have been running tests continuously for two weeks and the issue did not appear again, so from my point of view this fix is good.
Thanks again for providing so much insight besides fixing the underlying issue. I really appreciate that and I'll definitely give those things a closer look when time allows.

**Comment by giltene** [comment]
Great! 

With the fix confirmed, I'll likely release and push new version of HdrHistogram to maven central soon.

**Comment by rrakos-evo** [comment]
Thank you. 🙂

**Comment by rrakos-evo** [comment]
Could you please release a new version of HdrHistogram which we could use in Prometheus?

**Comment by giltene** [comment]
I'll likely get a release out on the weekend of Oct. 26th. A new PackedHistogram [with the potential for dramatically lower footprints through packed array representations for typically sparse histogram data]  (and hopefully a ConcurrentPackedHistogram) may or may not make it into that release, and I want to give it a chance to settle in the coming week.

**Comment by rrakos-evo** [comment]
I'm planning to finish https://github.com/prometheus/client_java/pull/484 around the end of the year. May I ask for an update on the release?

**Comment by giltene** [comment]
Released 2.1.12, which includes https://github.com/HdrHistogram/HdrHistogram/commit/753fe5c16ddb089b42a6ff920c06e50ead032945
